### PR TITLE
Abate spurious resize warnings in `MultiMarginLoss` on CUDA

### PR DESCRIPTION
**Pull request by eqy**
The current `multi_margin_loss_cuda_out` implementation has a few cases where an output is intended to have shape `{0}` before resizing but is improperly set before this step, triggering `UserWarning`s. This PR changes the initial `at::empty` call to avoid this warning and alters the current resizing logic to match that of the CPU version to avoid the warning when calling `at::sum_out`.

The original issue was reported from the user forum by @hadaev8 [here](https://discuss.pytorch.org/t/warning-spam-then-using-multimarginloss/147492).

CC @ptrblck

Original repro provided by @hadaev8:
```
import torch
from torch import nn
from torch.nn import functional as F

bs = 56
model = nn.Linear(128, 22).cuda()
loss = nn.MultiMarginLoss()
x = torch.rand((bs, 128)).cuda()
targets = torch.randint(22, (bs,)).cuda()
out = model(x)
print(targets.shape)
print(out.shape)
loss(out, targets)
```

### CODE REVIEW

File: aten/src/ATen/native/cuda/MultiMarginLoss.cu [hunk 1]
```diff
@@ -248,21 +250,25 @@ Tensor& multi_margin_loss_cuda_out(
...
         }
-        at::sum_out(out, tmp_output, /*dims=*/IntArrayRef{});
+        out = at::sum(tmp_output, /*dims=*/IntArrayRef{});
+        need_resize_ = true;
       }
     }
   });
 
   if (!out.is_alias_of(out_)) {
     out_.copy_(out);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/MultiMarginLoss.cu:260]: this doesn't seem to be the correct fix. `out_` at this point is the `nframe` size, and `copy_` fills all the unneeded values, and only `resize_` then saves the day. We should avoid this.

Comment (eqy) [reply aten/src/ATen/native/cuda/MultiMarginLoss.cu]: Agreed, was there a historical version why the CPU logic for resizing didn't work? Seems to work for the user-submitted case and in `test_scalar_check`.

Comment (ngimel) [reply aten/src/ATen/native/cuda/MultiMarginLoss.cu]: I don't know, sorry, I didn't look at the cpu implementation

**Comment by y0uCeF** [comment]
Hello, thanks for the solution. I also faced the same message recently that I had to grep -v those lines.
Is there a way to suppress the UserWarning directly from python?

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
@eqy can you please rebase to trigger CI?

**Comment by y0uCeF** [comment]
Hello @eqy 
On the latest pytorch 1.12.1 I still get resize warnings. However, compared with your reproduction code mine contains the backward() call

```
import torch
from torch import nn

bs = 56
model = nn.Linear(128, 22).cuda()
loss = nn.MultiMarginLoss()
x = torch.rand((bs, 128)).cuda()
targets = torch.randint(22, (bs,)).cuda()
out = model(x)
print(targets.shape)
print(out.shape)
l = loss(out, targets)
l.backward()
```
Output
```
torch.Size([56])
torch.Size([56, 22])
/home/user/.conda/envs/env/lib/python3.7/site-packages/torch/autograd/__init__.py:175: UserWarning: An output with one or more elements was resized since it had shape [], which does not match the required output shape [56, 22]. This behavior is deprecated, and in a future PyTorch release outputs will not be resized unless they have zero elements. You can explicitly reuse an out tensor t by resizing it, inplace, to zero elements with t.resize_(0). (Triggered internally at  /opt/conda/conda-bld/pytorch_1659484801627/work/aten/src/ATen/native/Resize.cpp:17.)
  allow_unreachable=True, accumulate_grad=True)  # Calls into the C++ engine to run the backward pass
```
Should I open a new issue?

**Comment by maxwell-aladago** [comment]
This issue still persist in 1.13 when backward is called.

**Comment by eqy** [comment]
Thanks for the reminders, opened #96382 to address this issue for backward

**Pull request closed without merging**